File: Data Structure/Lab 13/StackOperation.c
// Write a menu driven program to implement following operations on the Stack 
// using an Array 
//  PUSH, POP, DISPLAY 
//  PEEP, CHANGE

#include<stdio.h>
#include<stdlib.h>

#define n 5
int Stack[n],top = -1;

void PUSH(int x){
    if(top == n - 1){
        printf("Stack overflow!");
        return;
    }
    else{
        Stack[++top] = x;
    }
}

int POP(){
    if(top == -1){
        printf("Stack underflow!");
        return -1;
    }
    else{
        return Stack[top--];
    }
}

void DISPLAY(){
    if(top == -1){
        printf("Stack underflow!");
        return;
    }
    else{
        for(int i = top ; i >= 0 ; i--){
            printf("%d",Stack[i]);
        }
        printf("\n");
    }
}

int PEEP(int i){
    if(top - i + 1 < 0 || i <= 0){
        printf("Invalid Index");
        return -1;
    }
    else{
        return Stack[top - i + 1];
    }
}

void CHANGE(int i , int x){
    if(top - i + 1 < 0 || i <= 0){
        printf("Invalid index!");
        return;
    }
    else{
        Stack[top - i + 1] = x;
    }
}

void main() {
    int choice = 0, val, pos,i;

    while (1) {
        printf("\n--- Stack Operation Menu ---\n");
        printf("1-Push an element.\n");
        printf("2-Pop an element.\n");
        printf("3-Display stack.\n");
        printf("4-Peep an element.\n");
        printf("5-Change an element.\n");
        printf("6-Exit\n");

        printf("Enter choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter value:");
                scanf("%d",&val);
                PUSH(val);
                break;
            case 2:
                POP();
                break;
            case 3:
                DISPLAY();
                break;
            case 4:
                printf("Enter index: ");
                scanf("%d", &i);
                PEEP(i);
                break;
            case 5:
                printf("Enter index: ");
                scanf("%d", &i);
                printf("Enter value: ");
                scanf("%d", &val);
                CHANGE(i,val);
                break;
            case 6:
                printf("Exiting program.\n");
                exit(0);
            default:
                printf("Invalid choice.\n");
        }
    }
}

File: Data Structure(Java)/Lab 11/StackOperation.java
// Write a menu driven program to implement following operations on the Stack 
// using an Array 
//  PUSH, POP, DISPLAY 
//  PEEP, CHANGE

import java.util.Scanner;

public class StackOperation {
    static int[] S;
    static int top = -1;
    static int n;

    static void PUSH(int x) {
        if (top >= n - 1) {
            System.out.println("Stack Overflow!");
        } else {
            S[++top] = x;
            System.out.println(x + " pushed into stack.");
        }
    }

    static int POP() {
        if (top == -1) {
            System.out.println("Stack Underflow!!");
            return -1;
        } else {
            return S[top--];
        }
    }

    static void DISPLAY() {
        if (top == -1) {
            System.out.println("Stack is empty!");
            return;
        }
        System.out.print("Stack elements: ");
        for (int i = top; i >= 0; i--) {
            System.out.print(S[i] + " ");
        }
        System.out.println();
    }

    static int PEEP(int i) {
        // i represents the ith element from the top (1st from top, 2nd from top, etc.)
        if (top - i + 1 < 0 || i <= 0) {
            System.out.println("Invalid Index");
            return -1;
        } else {
            return S[top - i + 1];
        }
    }

    static void CHANGE(int i, int x) {
        if (top - i + 1 < 0 || i <= 0) {
            System.out.println("Invalid Index!!");
        } else {
            S[top - i + 1] = x;
            System.out.println("Value changed successfully.");
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        // CRITICAL FIX: Initialize stack size and array
        System.out.print("Enter the size of the stack: ");
        n = sc.nextInt();
        S = new int[n];

        boolean run = true;
        while (run) {
            System.out.println("\n--- Stack Operations Menu ---");
            System.out.println("1 - PUSH");
            System.out.println("2 - POP");
            System.out.println("3 - DISPLAY");
            System.out.println("4 - PEEP");
            System.out.println("5 - CHANGE");
            System.out.println("6 - Exit");
            System.out.print("Enter your choice (1-6): ");
            
            int choice = sc.nextInt();
            switch (choice) {
                case 1:
                    System.out.print("Enter value to insert: ");
                    int val = sc.nextInt();
                    PUSH(val);
                    break;
                case 2:
                    int poppedVal = POP();
                    if (poppedVal != -1) {
                        System.out.println("Popped value: " + poppedVal);
                    }
                    break;
                case 3:
                    DISPLAY();
                    break;
                case 4:
                    System.out.print("Enter position from top (1-based): ");
                    int idx = sc.nextInt();
                    int peepedVal = PEEP(idx);
                    if (peepedVal != -1) {
                        System.out.println("Value at position " + idx + " from top is: " + peepedVal);
                    }
                    break;
                case 5:
                    System.out.print("Enter position from top to change (1-based): ");
                    int j = sc.nextInt();
                    System.out.print("Enter new value: ");
                    int x = sc.nextInt();
                    CHANGE(j, x);
                    break;
                case 6:
                    System.out.println("Exiting program.");
                    run = false;
                    break;
                default:
                    System.out.println("Invalid choice! Please select between 1-6.");
                    break;
            }
        }
        sc.close();
    }
}
